instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go

```diff
@@ -0,0 +1,114 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelaws
+
+import (
+	"context"
+
+	v2Middleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/semconv"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go-v2/otelaws"
+)
+
+type oTelMiddlewares struct {
+	tracer trace.Tracer
+}
+
+func (m oTelMiddlewares) spanCreator(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("OtelSpanCreator", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error) {
+
+		opts := []trace.SpanOption{
+			trace.WithSpanKind(trace.SpanKindClient),
+		}
+		ctx, span := m.tracer.Start(ctx, "AWS", opts...)
+		defer span.End()
+
+		out, metadata, err = next.HandleInitialize(ctx, in)
+		if err != nil {
+			span.RecordError(err)
+		}
+
+		return out, metadata, err
+	}),
+		middleware.Before)
+}
+
+func (m oTelMiddlewares) spanNameSetter(stack *middleware.Stack) error {
+	return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("OtelSpanNameSetter", func(
+		ctx context.Context, in middleware.InitializeInput, next middleware.InitializeHandler) (
+		out middleware.InitializeOutput, metadata middleware.Metadata, err error) {
+
+		span := trace.SpanFromContext(ctx)
+		span.SetName(v2Middleware.GetServiceID(ctx))
```

Comment:
Also, you can bring the following lines out of Deserialize to here too as these will have been set by Init After.
```go
span.SetAttributes(ServiceAttr(v2Middleware.GetServiceID(ctx)),
			RegionAttr(v2Middleware.GetRegion(ctx)),
			OperationAttr(v2Middleware.GetOperationName(ctx)))
```

Reply:
Nice catch.